Create API for easier integration with 3rd party apps

timrae opened the pull request:
Here's the overdue work that I've been planning on changing the content provider and writing a higher-level API. There are quite a lot of changes... Any volunteers to give it a proper code review? @flerda ;)

Here are a few of the major changes that I remember making:
* Added a high level API for adding cards to AnkiDroid from external apps. I plan to release the API under the LGPL license to allow commercial apps to use it. I'll release the API to maven after this is merged so that 3rd party apps can use the API by just including it in their build.gradle files.
* Removed the `notes_data` URI from the ContentProvider in favor of using the `flds` encoded in the same way as the DB. I've explained the reasons for this [elsewhere](https://groups.google.com/d/topic/anki-android/QCv1Yzxmzj0/discussion)
* Decks are now added via a separate interface from notes, as per above link
* Added some extra URIs to the provider for inserting and updating models / templates, deleting notes (in order to enable previewing of cards before they are added), querying the current model, etc.
* Changed the unit tests around a bit

I've also created a sample app that shows how to use the API. I'll release that to a separate project on our github after I've released the API to maven. In the meantime you can see it [here](https://github.com/timrae/Anki-Android/tree/api/apisample/src/main) if you like.

I have updated the existing tests for the ContentProvider, and added a couple more for the new code, but I think there might still be some missing. Also, we really need to add a unit test for each method of the public API. I haven't really tested the preview functionality in the API yet.

flerda commented:
Sure! I will have a proper look when I have some free time.

flerda commented on AnkiDroid/src/androidTest/java/com.ichi2.anki.tests/ContentProviderTest.java:
```diff
@@ -70,124 +80,71 @@
+        // Delete all notes
+        int deletedNotes = 0;
+        for (Uri uri: mCreatedNotes) {
+            final ContentResolver cr = getContext().getContentResolver();
```
This can be done outside the loop.

flerda commented on AnkiDroid/src/androidTest/java/com.ichi2.anki.tests/ContentProviderTest.java:
```diff
@@ -70,124 +80,71 @@
+        int deletedNotes = 0;
+        for (Uri uri: mCreatedNotes) {
+            final ContentResolver cr = getContext().getContentResolver();
+            cr.delete(uri, null, null);
```
Would it make sense to do the delete using operations on ```col``` rather than using the content provider?

This is a test for the content provider itself and its ```tearDown()``` operating successfully depends on the correctness of the content provider right now.

timrae replied:
Yeah you're right, tbh I was just being lazy not wanting to add extra tests lol

flerda commented on AnkiDroid/src/androidTest/java/com.ichi2.anki.tests/ContentProviderTest.java:
```diff
@@ -70,124 +80,71 @@
-                assertEquals("Tag set", 1, cr.update(newNoteDataUri, values, null, null));
-                mCreatedNotes++;
+        mCreatedNotes = new ArrayList<>();
+        ProviderUtils api = new ProviderUtils(getContext());
```
This means the ```setup()``` method depends on the content provider implementation, which is not great.
Only the test code itself should depend on the content provider.

flerda commented on AnkiDroid/src/androidTest/java/com.ichi2.anki.tests/ContentProviderTest.java:
```diff
@@ -255,151 +212,160 @@ public void testQueryNotesProjection() {
+            Cursor noteCursor = cr.query(uri, FlashCardsContract.Note.DEFAULT_PROJECTION, null, null, null);
+            try {
+                assertNotNull("Check that there is a valid cursor for detail data after update", noteCursor);
+                assertEquals("Check that there is one any only one entry after update", 1, noteCursor.getCount());
```
one any only one -> one and only one

flerda commented on AnkiDroid/src/androidTest/java/com.ichi2.anki.tests/ContentProviderTest.java:
```diff
@@ -255,151 +212,160 @@ public void testQueryNotesProjection() {
-                // After update query again
-                noteDataCursor = cr.query(dataUri, null, null, null, null);
-                assertNotNull("Check that there is a valid cursor for detail data after update", noteDataCursor);
+            } catch (JSONException e) {
```
It is generally considered fine to throw exceptions in tests. In fact, the code as written obfuscates the failure because it hides the exact stack trace that caused it.

I would suggested declaring the test method as:
```java
  public void testInsertAndUpdateModel() throws Exception {
    ...
  }
```
without the inner try-catch, so that failures can be more easily diagnosed.

flerda commented on AnkiDroid/src/androidTest/java/com.ichi2.anki.tests/ContentProviderTest.java:
```diff
@@ -255,151 +212,160 @@ public void testQueryNotesProjection() {
             }
         } finally {
-            allNotesCursor.close();
+            // Delete the model (this will cause a full-sync so don't run the tests on your main device)
```
This line is not very helpful here, since it is unlikely anyone will ever read it when running the tests. Maybe put a similar comment in the development guide where we talk about running tests.

timrae replied:
Good point! I added a comment to the Wiki

flerda commented on AnkiDroid/src/androidTest/java/com.ichi2.anki.tests/ContentProviderTest.java:
```diff
@@ -410,21 +376,35 @@ public void testMoveToOtherDeck() {
         }
     }
 
+
```
For consistency, add a Javadoc comment here as well.

flerda commented on AnkiDroid/src/androidTest/java/com.ichi2.anki.tests/ContentProviderTest.java:
```diff
@@ -532,7 +502,7 @@ public void testQueryCertainDeck(){
         col = CollectionHelper.getInstance().getCol(getContext());
 
-        long deckId = testDeckIds[0]; //<-- insert real deck ID here
+        long deckId = mTestDeckIds[0]; //<-- insert real deck ID here
```
Is the ```//<-- insert real deck ID here``` command actually incorrect, since we are already putting the deck ID we obtained when creating the test deck?

flerda commented on AnkiDroid/src/main/java/com/ichi2/anki/provider/CardContentProvider.java:
```diff
@@ -51,20 +56,22 @@
  * .../notes/#/cards/# (access specific card of note)
  * .../models (search for models)
  * .../models/# (direct access to model)
+ * .../current_model (direct access to the currently active model)
```
Suggestion.

How about using:
```
.../models/current
```
instead of:
```
.../current_model
```

It seems to better mimic:
```
.../models/#
```
which should offer the same structure.

timrae replied:
I tried that but it seemed it was not possible to add both `.../models/#` and `.../models/current` to the URIMatcher. The closest option would probably be to have a single `.../models/*` and then parse the argument. Do you think this would be useful?

flerda commented on AnkiDroid/src/main/java/com/ichi2/anki/provider/CardContentProvider.java:
```diff
@@ -51,20 +56,22 @@
  * .../models/# (direct access to model)
+ * .../current_model (direct access to the currently active model)
+ * .../models/#/fields (access to field definitions of a model)
+ * .../models/#/templates (access to card templates of a model)
```
Do we want to support ```.../models/current/fields``` and ```.../models/current/templates``` as well?

timrae replied:
That could potentially be useful, but the api currently only gives back the model ID for the current model, so probably no need to go out my way to add support for it just yet

flerda commented on AnkiDroid/src/main/java/com/ichi2/anki/provider/CardContentProvider.java:
```diff
@@ -72,11 +79,13 @@
     private static final int NOTES_ID = 1001;
-    private static final int NOTES_ID_DATA = 1002;
-    private static final int NOTES_ID_CARDS = 1003;
+    private static final int NOTES_ID_CARDS = 1003; // 1002 is reserved (from early alpha version)
```
These constants are internal only, so you can actually change them (and reuse them) without affecting previous version of the app.

flerda commented on AnkiDroid/src/main/java/com/ichi2/anki/provider/CardContentProvider.java:
```diff
@@ -72,11 +79,13 @@
-    private static final int NOTES_ID_DATA = 1002;
-    private static final int NOTES_ID_CARDS = 1003;
+    private static final int NOTES_ID_CARDS = 1003; // 1002 is reserved (from early alpha version)
     private static final int NOTES_ID_CARDS_ORD = 1004;
```
Why not ```NOTES_ID_CARDS_ID```? It seems it would better match the naming of other constants (ID is basically representing the ```#``` in the template).

timrae replied:
They're slightly different as ID refers to the actual ID in the anki database, whereas ORD goes from 0 .. numCards -1

flerda commented on AnkiDroid/src/main/java/com/ichi2/anki/provider/CardContentProvider.java:
```diff
@@ -288,30 +248,64 @@ public Cursor query(Uri uri, String[] projection, String selection,
+                 */
+                int ord = Integer.parseInt(uri.getLastPathSegment());
+                JSONObject currentModel = getModelFromUri(uri, col);
+                MatrixCursor rv;
```
You can actually initialize rv when declared instead by moving this line down.

flerda commented on AnkiDroid/src/main/java/com/ichi2/anki/provider/CardContentProvider.java:
```diff
@@ -580,7 +645,26 @@ public int update(Uri uri, ContentValues values, String selection,
+
+        int updated = 0; // Number of updated entries (return value)
+        switch (match) {
+            case NOTES_ID:
```
Use ```{ }``` to enclose a case block that declares variables to assure a proper scope for them.

```java
  case NOTES_ID: {
    long nodeId = ...
    ...
  }
```

flerda commented on AnkiDroid/src/main/java/com/ichi2/anki/provider/CardContentProvider.java:
```diff
@@ -580,7 +645,26 @@ public int update(Uri uri, ContentValues values, String selection,
+            case NOTES_ID:
+                long noteId = Long.parseLong(uri.getPathSegments().get(1));
+                col.remNotes(new long[] { noteId });
+                updated++;
```
I would get rid of the ```updated``` variable and just use here:
```java
return 1;
```

flerda commented on api/src/main/java/com/ichi2/anki/api/ProviderUtils.java:
```diff
@@ -0,0 +1,375 @@
+ */
+public final class ProviderUtils {
+    private final ContentResolver mResolver;
+    final String TEST_TAG = "PREVIEW_NOTE";
```
Can this be private?

flerda commented on api/src/main/java/com/ichi2/anki/api/ProviderUtils.java:
```diff
@@ -0,0 +1,375 @@
+        return resolver.acquireContentProviderClient(FlashCardsContract.AUTHORITY_URI) != null;
+    }
+
+    public static String getPackageName(Context context) {
```
What is the point of this method from the point of view of an API users?

timrae replied:
It's so that they can know which app the provider belongs to. In the sample app I used it [like this](https://github.com/timrae/Anki-Android/blob/api/apisample/src/main/java/com/ichi2/apisample/MainActivity.java#L183) to get the name and icon. It seemed better than getting the 3rd party apps to hardcode in "com.ichi2.anki", as there may be forks, or we may want to take advantage of flavors etc in the future

flerda commented on api/src/main/java/com/ichi2/anki/api/ProviderUtils.java:
```diff
@@ -0,0 +1,375 @@
+
+    public static String getPackageName(Context context) {
+        PackageManager manager = context.getPackageManager();
+        return manager.resolveContentProvider(FlashCardsContract.AUTHORITY, 0).packageName;
```
This will crash if the app is not installed.
Maybe it would be better to define this to return null instead in that case?

flerda commented on api/src/main/java/com/ichi2/anki/api/ProviderUtils.java:
```diff
@@ -0,0 +1,375 @@
+
+    public static boolean isInstalled(Context context) {
+        ContentResolver resolver = context.getApplicationContext().getContentResolver();
+        return resolver.acquireContentProviderClient(FlashCardsContract.AUTHORITY_URI) != null;
```
You must call ```release()``` on the acquired content provide if it is not null.
See:
http://developer.android.com/reference/android/content/ContentResolver.html#acquireContentProviderClient(java.lang.String)

flerda commented on settings.gradle:
```diff
@@ -1,2 +1,2 @@
-include ':ShowcaseView:library'
```
Why is this deleting ShowcaseView?

timrae replied:
ShowcaseView was deleted a while ago, I had just forgotten to remove it from settings.gradle

flerda commented on api/src/main/java/com/ichi2/anki/api/ProviderUtils.java:
```diff
@@ -0,0 +1,375 @@
+/***************************************************************************************
```
I like what this is doing, basically providing a client library that third-party apps can build against.

One question I would have is whether the client library can be the only API that a third-party needs to worry about, without having to care about the underlying content provider.

What I mean is that ```FlashCardsContract``` could be made package-private (so that client code cannot access it) and ```ProviderUtils``` could be renamed to something more powerful, e.g., ```AnkiDroidApi```.

Then, client code would look like:
```java
public void addFlashCardForWord(DictEntry entry) {
  AnkiDroidApi ankiDroidApi = new AnkiDroidApi(getContext());
  long modelId = ankiDroidApi.findModel("com.mydict.chinese");
  if (modelId == 0) {
    modelId = ankiDroidApi.addNewModel("com.mydict.chinese", ...);
  }
  ankiDroidApi.addNewNote(modelId, ...);
}
```
without having to actually need to know about the complexity of the content provider.

What do you think?

timrae replied:
> One question I would have is whether the client library can be the only API that a third-party needs to worry about, without having to care about the underlying content provider.

Well, it's already true that a third-party doesn't need to care about the underlying provider, and they could use that simple pattern to add cards (though I think it's better to [use the mid like I did in the sample](https://github.com/timrae/Anki-Android/blob/api/apisample/src/main/java/com/ichi2/apisample/MainActivity.java#L231) since model names are not unique in libanki, and they can be renamed by the user)

Renaming the class is a good idea, I'll do that! I had thought that `com.ichi2.anki.api.Api` sounded a bit stupid, but your idea of `com.ichi2.anki.api.AnkiDroidApi` sounds better, so maybe we can go with that.

Making the provider contract package private has the disadvantage that we would then need to maintain 2 copies of the file (one in the main package and one in the api package), which I was trying to avoid. There is also functionality in the provider not currently exposed in the API, such as the schedule table, which is used in the smartwatch app.

If we wanted to go out of our way to hide the details, we could put the necessary missing stuff into the API, and move the contract into a separate library. I'm not sure if this would achieve much at the end of the day though, since the provider is open, and if someone really wants to use it directly they could.

I'm planning to write some proper documentation on the wiki, so how about we just don't document the contract file, and in the main api doc we can write something like "Note: the `AnkiDroidApi` class internally uses a `ContentProvider`, and while it's possible access the provider directly, this is not recommended as it could be subject to breaking changes in the future."...?

timrae commented:
Thanks for the detailed review, I've implemented most of the changes discussed, except for decoupling the `setup()` method from the ContentProvider, where I've simply added a comment that we should fix that up at some later point. As far as I can tell, it's very unlikely for the current way to make the tests NOT fail when they're supposed to, so it would only serve to help us more quickly pinpoint the root-cause of failures when they do occur.

timrae commented:
Btw, I squashed the commits down to one, but for reference [here is the unsquashed version](https://github.com/timrae/Anki-Android/commits/api2). Please let me know if you have any further comments, otherwise I'll go ahead and merge it :)

timrae commented:
I started writing the documentation:
https://github.com/ankidroid/Anki-Android/wiki/AnkiDroid-API-(alpha)

timrae commented:
@flerda I liked the simplicity of the example code you were imagining, so I moved the storage of the did and mid in `SharedPreferences` into the API itself. I also added two convenience methods for adding basic / basic + reverse models, without requiring the app to define a custom model.

Those changes clean it up quite nicely I think. I think it's basically ready now so will go ahead and merge. I would like to add duplicate check functionality to the API relatively soon as well.

timrae commented:
oops, just realized that adding the Context to the new methods was unnecessary; we can just keep a reference to the `ApplicationContext` in the constructor.
